openedx/features/enterprise_support/api.py

```diff
@@ -592,109 +592,129 @@ def enterprise_customer_for_request(request):
     return enterprise_customer
 
 
+def get_data_sharing_consent_info(username, enterprise_customer_uuid, course_key, enrollment_exists):
+    """
+    Return data sharing consent record for an enterprise learner.
+    """
+    # get the proxy DSC record if it exists
+    dsc = get_course_data_sharing_consent(
+        username,
+        course_key,
+        enterprise_customer_uuid,
+    )
+
+    if dsc:
+        dsc_info = dsc.serialize()
+
+        LOGGER.info("[ENTERPRISE DSC] Learner DSC Details: [{}]".format(dsc_info))
+        # Learner is already enrolled in a course but no enterprise record exists so consent is not needed.
+        if enrollment_exists and not dsc_info.get('exists', False):
+            return False
+
+        return dsc_info['consent_required']
+
+
+def _has_learner_already_given_consent(existing_dsc_records, learner_consent_input_data):
+    """
+    Return True if a learner has already given the consent for a course and an enterprise, False otherwise.
+    """
+    for record in existing_dsc_records:
+        if record['consent_provided']:
+            record_consent_data = (record['username'], str(record['enterprise_customer_uuid']), record['course_id'])
+            if record_consent_data == learner_consent_input_data:
+                return True
+    return False
+
+
 @enterprise_is_enabled(otherwise=False)
 def consent_needed_for_course(request, user, course_id, enrollment_exists=False):
     """
     Wrap the enterprise app check to determine if the user needs to grant
     data sharing permissions before accessing a course.
     """
     LOGGER.info(
-        "Determining if user [{username}] must consent to data sharing for course [{course_id}]".format(
+        "[ENTERPRISE DSC] Determining if user [{username}] must consent to data sharing for course"
+        " [{course_id}]".format(
             username=user.username,
             course_id=course_id
         )
     )
 
-    enterprise_learner_details = get_enterprise_learner_data_from_db(user)
-    enterprise_customer_uuid = None
-    if enterprise_learner_details:
-        enterprise_customer = enterprise_learner_details[0]['enterprise_customer']
-        enterprise_customer_uuid = enterprise_customer['uuid']
-
-    consent_cache_key = get_data_consent_share_cache_key(user.id, course_id, enterprise_customer_uuid)
```

Comment:
We do not want to remove Caching we are calling this method for every user learning MFE.